lib/vector-core/src/stream/batcher.rs

```diff
@@ -0,0 +1,538 @@
+            batches: HashMap::default(),
+            closed_batches: Vec::default(),
+            timer,
+            partitioner,
+            stream,
+        }
+    }
+}
+
+impl<St, Prt, T> Stream for Batcher<St, Prt, T>
+where
+    St: Stream + Unpin,
+    Prt: Partitioner + Unpin,
+    Prt::Item: ByteSizeOf + From<St::Item>,
+    T: Timer,
+{
+    type Item = (Option<Prt::Key>, Vec<Prt::Item>);
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.stream.size_hint()
+    }
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut this = self.project();
+        loop {
+            if !this.closed_batches.is_empty() {
+                return Poll::Ready(this.closed_batches.pop());
+            }
+            match this.stream.as_mut().poll_next(cx) {
+                Poll::Pending => {
+                    if this.timer.has_elapsed() {
+                        this.closed_batches.extend(
+                            this.batches
+                                .drain()
+                                .map(|(key, batch)| (key, batch.destruct().1)),
+                        );
+                        this.timer.reset();
+                    }
+                    return Poll::Pending;
+                }
+                Poll::Ready(None) => {
+                    // Now that the underlying stream is closed we need to clear
+                    // out our batches.
+                    if !this.batches.is_empty() {
+                        this.closed_batches.extend(
+                            this.batches
+                                .drain()
+                                .map(|(key, batch)| (key, batch.destruct().1)),
+                        );
+                        continue;
+                    }
+                    return Poll::Ready(None);
+                }
+                Poll::Ready(Some(item)) => {
+                    let item: Prt::Item = item.into();
+                    let item_key = this.partitioner.partition(&item);
+
+                    if let Some(batch) = this.batches.get_mut(&item_key) {
+                        if batch.has_space(&item) {
+                            batch.push(item);
+                        } else {
+                            let new_batch =
+                                Batch::new(*this.batch_item_limit, *this.batch_allocation_limit)
+                                    .with(item);
+                            let batch = mem::replace(batch, new_batch);
+                            let (_, arr) = batch.destruct();
+                            if this.timer.has_elapsed() {
+                                this.closed_batches.extend(
+                                    this.batches
+                                        .drain()
+                                        .map(|(key, batch)| (key, batch.destruct().1)),
+                                );
+                                this.timer.reset();
```

Comment:
Do we think this will provide a meaningful throughput or latency improvement vs just letting the loop go back around to hit the existing "has timer elapsed?" logic?

Reply:
With regard to just the reset? If the broader code block, after a batch is issued I wanted to check if the timer elapsed to flush any slow partitions. If just with regard to the reset I wanted to avoid a situation were we issue not-full batches if it can be helped. The timeout triggering forces some not-full batches to issue, of course.

Reply:
Makes sense, yeah.  I was really only thinking about it from the code duplication standpoint, but it's innocuous enough.